File: shoebill_plugins/RadioPlugin/src/main/java/lt/ltrp/RadioPlugin.java
package lt.ltrp;

import lt.ltrp.dao.RadioStationDao;
import lt.ltrp.dao.impl.SqlRadioStationDao;
import net.gtaun.shoebill.event.resource.ResourceLoadEvent;
import net.gtaun.shoebill.resource.Plugin;

import javax.sql.DataSource;

/**
 * @author Bebras
 *         2016.04.13.
 */
public class RadioPlugin extends Plugin implements RadioController {

    private RadioStationDao dao;

    @Override
    protected void onEnable() throws Throwable {
        Instance.instance = this;
        getEventManager().registerHandler(ResourceLoadEvent.class,  e-> {
            if(e.getResource().getClass().equals(DatabasePlugin.class)) {
                DataSource ds = ((DatabasePlugin)e.getResource()).getDataSource();
                this.dao = new SqlRadioStationDao(ds);
            }
        });
    }

    @Override
    protected void onDisable() throws Throwable {

    }



    public RadioStationDao getRadioStationDao() {
        return dao;
    }
}

File: shoebill_plugins/RadioPlugin/src/main/java/lt/ltrp/radio/RadioPlugin.kt
package lt.ltrp.radio

import lt.ltrp.DatabasePlugin
import lt.ltrp.radio.dao.RadioStationDao
import lt.ltrp.radio.dao.impl.SqlRadioStationDao
import lt.ltrp.radio.data.RadioStation
import lt.ltrp.resource.DependentPlugin
import net.gtaun.shoebill.ShoebillMain
import net.gtaun.shoebill.resource.ResourceManager
import net.gtaun.util.event.EventManagerNode


/**
 * @author Bebras
 *         2016.04.13.
 *
 *         Radio station management plugin, provides abstract classes for implementing an in-game radio
 *
 *         Future plans:
 *         <ul>
 *            <li>Dialogs for CRUD</li>
 *         </ul>
 */
@ShoebillMain("LTRP Radio plugin", "Bebras")
class RadioPlugin : DependentPlugin() {

    private lateinit var eventManagerNode: EventManagerNode
    lateinit var stationDao: RadioStationDao
        private set

    init {
        addDependency(DatabasePlugin::class)
    }

    override fun onEnable() {
        eventManagerNode = eventManager.createChildNode()
    }

    override fun onDisable()  {
        eventManagerNode.cancelAll()
    }

    override fun onDependenciesLoaded() {
        val ds = ResourceManager.get().getPlugin(DatabasePlugin::class.java)?.dataSource
        this.stationDao = SqlRadioStationDao(ds, logger)
        RadioStation.stations = stationDao.get()
        logger.info("Loaded " + RadioStation.stations.size + " radio stations")
    }

}
